File: smoke_tests/lua/xss.lua
-- Cross-Site Scripting (XSS) vulnerabilities in Lua
-- OpenResty/nginx-lua patterns

-- Test 1: Direct output of user input
function echo_input()
    local args = ngx.req.get_uri_args()
    local name = args.name
    -- VULNERABLE: Reflected XSS
    ngx.say("<h1>Hello, " .. name .. "!</h1>")
end

-- Test 2: ngx.print without escaping
function print_message()
    local args = ngx.req.get_uri_args()
    local msg = args.message
    -- VULNERABLE: XSS via ngx.print
    ngx.print("<div class='message'>" .. msg .. "</div>")
end

-- Test 3: Header injection
function set_custom_header()
    local args = ngx.req.get_uri_args()
    local value = args.header_value
    -- VULNERABLE: Header injection
    ngx.header["X-Custom"] = value
end

-- Test 4: Cookie value from user input
function set_cookie()
    local args = ngx.req.get_uri_args()
    local token = args.token
    -- VULNERABLE: Cookie injection
    ngx.header["Set-Cookie"] = "session=" .. token
end

-- Test 5: HTML template injection
function render_template(user_data)
    -- VULNERABLE: Template injection
    local html = [[
        <html>
        <body>
            <h1>Welcome, ]] .. user_data.name .. [[!</h1>
            <p>Email: ]] .. user_data.email .. [[</p>
        </body>
        </html>
    ]]
    ngx.say(html)
end

-- Test 6: JSON response with user data
function json_response()
    local args = ngx.req.get_uri_args()
    local callback = args.callback
    local data = '{"status": "ok"}'
    -- VULNERABLE: JSONP callback injection
    ngx.say(callback .. "(" .. data .. ")")
end

-- Test 7: Error message with user input
function show_error()
    local args = ngx.req.get_uri_args()
    local error_msg = args.error
    -- VULNERABLE: Error message XSS
    ngx.say("<div class='error'>Error: " .. error_msg .. "</div>")
end

-- Test 8: URL in redirect
function redirect_user()
    local args = ngx.req.get_uri_args()
    local next_url = args.next
    -- VULNERABLE: Open redirect and potential XSS
    ngx.redirect(next_url)
end

-- Test 9: Building HTML table from data
function render_table(rows)
    ngx.say("<table>")
    for _, row in ipairs(rows) do
        -- VULNERABLE: XSS in table cells
        ngx.say("<tr><td>" .. row.name .. "</td><td>" .. row.value .. "</td></tr>")
    end
    ngx.say("</table>")
end

-- Test 10: JavaScript variable injection
function inject_js_var()
    local args = ngx.req.get_uri_args()
    local user_id = args.user_id
    -- VULNERABLE: JavaScript injection
    ngx.say("<script>var userId = '" .. user_id .. "';</script>")
end

-- Test 11: Attribute injection
function render_link()
    local args = ngx.req.get_uri_args()
    local url = args.url
    local text = args.text
    -- VULNERABLE: Attribute injection
    ngx.say('<a href="' .. url .. '">' .. text .. '</a>')
end

-- Test 12: Style injection
function custom_style()
    local args = ngx.req.get_uri_args()
    local color = args.color
    -- VULNERABLE: CSS injection
    ngx.say('<div style="color: ' .. color .. '">Styled text</div>')
end

-- Test 13: Event handler injection
function render_button()
    local args = ngx.req.get_uri_args()
    local action = args.action
    -- VULNERABLE: Event handler XSS
    ngx.say('<button onclick="' .. action .. '">Click me</button>')
end

File: smoke_tests/perl/xss.pl
#!/usr/bin/perl
# Cross-Site Scripting (XSS) vulnerabilities in Perl CGI

use strict;
use warnings;
use CGI;

my $cgi = CGI->new();

# Test 1: Direct print of user input
sub vulnerable_print {
    my $name = $cgi->param('name');
    print "Content-Type: text/html\n\n";
    # VULNERABLE: Reflected XSS
    print "<h1>Hello, $name!</h1>";
}

# Test 2: say with user input
sub vulnerable_say {
    my $message = $cgi->param('message');
    print "Content-Type: text/html\n\n";
    # VULNERABLE: XSS via say
    say "<div>$message</div>";
}

# Test 3: printf with user input
sub vulnerable_printf {
    my $username = $cgi->param('username');
    print "Content-Type: text/html\n\n";
    # VULNERABLE: XSS via printf
    printf("<span class='user'>%s</span>", $username);
}

# Test 4: CGI start_html with user input
sub vulnerable_start_html {
    my $title = $cgi->param('title');
    # VULNERABLE: XSS in page title
    print $cgi->header();
    print $cgi->start_html(-title => $title);
    print $cgi->end_html();
}

# Test 5: CGI hidden field
sub vulnerable_hidden_field {
    my $value = $cgi->param('value');
    print $cgi->header();
    print $cgi->start_html();
    # VULNERABLE: XSS in hidden field
    print $cgi->hidden(-name => 'data', -value => $value);
    print $cgi->end_html();
}

# Test 6: CGI textfield
sub vulnerable_textfield {
    my $default = $cgi->param('default');
    print $cgi->header();
    print $cgi->start_html();
    # VULNERABLE: XSS in form field default
    print $cgi->textfield(-name => 'input', -default => $default);
    print $cgi->end_html();
}

# Test 7: HTML::Template with unescaped variable
sub vulnerable_html_template {
    use HTML::Template;
    my $name = $cgi->param('name');
    my $template = HTML::Template->new(filename => 'template.tmpl');
    # VULNERABLE: Template variable not escaped
    $template->param(NAME => $name);
    print $cgi->header();
    print $template->output();
}

# Test 8: Template Toolkit without AUTO_FILTER
sub vulnerable_template_toolkit {
    use Template;
    my $comment = $cgi->param('comment');
    my $tt = Template->new();
    # VULNERABLE: No auto-escaping
    $tt->process('page.tt', { comment => $comment });
}

# Test 9: Here-doc with interpolation
sub vulnerable_heredoc {
    my $content = $cgi->param('content');
    print "Content-Type: text/html\n\n";
    # VULNERABLE: XSS via here-doc
    print <<"HTML";
<html>
<body>
<div class="content">$content</div>
</body>
</html>
HTML
}

# Test 10: JSON output with user data
sub vulnerable_json_output {
    my $callback = $cgi->param('callback');
    my $data = '{"status":"ok"}';
    print "Content-Type: application/javascript\n\n";
    # VULNERABLE: JSONP callback injection
    print "$callback($data)";
}

# Test 11: Error message reflection
sub vulnerable_error_message {
    my $input = $cgi->param('input');
    if (!$input) {
        print $cgi->header();
        print $cgi->start_html();
        # VULNERABLE: Error message XSS
        print "<p class='error'>Invalid input: $input</p>";
        print $cgi->end_html();
    }
}

# Test 12: URL parameter in link
sub vulnerable_link {
    my $url = $cgi->param('url');
    print $cgi->header();
    print $cgi->start_html();
    # VULNERABLE: XSS via javascript: URL
    print "<a href='$url'>Click here</a>";
    print $cgi->end_html();
}

# Test 13: Event handler injection
sub vulnerable_event_handler {
    my $action = $cgi->param('action');
    print $cgi->header();
    print $cgi->start_html();
    # VULNERABLE: Event handler XSS
    print "<button onclick='$action'>Submit</button>";
    print $cgi->end_html();
}

# Test 14: Style attribute injection
sub vulnerable_style {
    my $color = $cgi->param('color');
    print $cgi->header();
    print $cgi->start_html();
    # VULNERABLE: CSS injection
    print "<div style='background-color: $color'>Content</div>";
    print $cgi->end_html();
}

# Test 15: Cookie value reflection
sub vulnerable_cookie_reflection {
    my $cookie_value = $cgi->cookie('user_pref');
    print $cgi->header();
    print $cgi->start_html();
    # VULNERABLE: Cookie-based XSS
    print "<div>Your preference: $cookie_value</div>";
    print $cgi->end_html();
}

1;
